### ISSUE TITLE
Add support for Trimming/AOT/SingleFile

### ISSUE THREAD

**Issue by lilith** [xoofx/Tomlyn#80]
While certain runtime serializations are pretty hard to do under AOT/Trim, I think most functionality could remain with it enabled, and the rest could work via a source analyzer. Either way, it would be good to annotate the source code into the AOT-compatible and not portions. Drop these into Tomlyn.props (and consider adding a net8.0) target to see the warnings - there aren't many, but they may be essential.  

```
    <EnableTrimAnalyzer>true</EnableTrimAnalyzer>
    <EnableSingleFileAnalyzer>true</EnableSingleFileAnalyzer>
    <EnableAotAnalyzer>true</EnableAotAnalyzer>
```

I went about the codebase adding `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.PublicMethods)]` to Types passed around, but I think there might need to be a little refactoring to prevent unlimited spread of those annotations.

There's also another option - System.Text.Json already supports source generation and allows turning a JsonNode (tree) into a custom object. I'm unsure how exact the mapping between toml and json is, but converting the parsed toml tree to a jsonnode could be a workaround for some.

**Comment by lilith** [comment]
I would say that the first step would be adding a .NET 8 target across the projects and to CI, then adding some failing tests that explore exactly what is currently broken under trimming. Would you want to keep .NET 7 or replace it with .NET 6? I see lots of .NET 6 ifdefs. .NET 6 EOL is November 12, 2024, [.NET 7 EOL is May 14, 2024](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core).

I don't have deep enough knowledge of the code architecture to refactor what is needed for direct custom model support - the design seems to be partly supportive of the changes but I would likely take it an unhelpful direction without knowing your goals/philosophy. I also don't have that bandwidth, and while my new docker images currently use TOML, I could see switching to JSON as being less costly. 

Adding JsonNode translation would introduce a new dependency (although perhaps not on .NET 8?), but could be isolated in code path to allow it to be trimmed out. I'd either do a separate package or make the feature .NET 8 only.

**Comment by lilith** [comment]
I opened a separate issue for .NET 8: https://github.com/xoofx/Tomlyn/issues/82

**Comment by xoofx** [comment]
> I went about the codebase adding [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.PublicMethods)] to Types passed around, but I think there might need to be a little refactoring to prevent unlimited spread of those annotations.

I dunno how much trouble it will be for Tomlyn, as I forgot the details about the codebase. I tried to do similar stuffs in another OSS project [Scriban](https://github.com/scriban/scriban) that is heavily relying on reflection, but it was too laborious to get it working.

> Drop these into Tomlyn.props (and consider adding a net8.0) target to see the warnings - there aren't many, but they may be essential.

For .NET 8, I have been using the following (as described [here](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows#aot-compatibility-analyzers))

```xml
<PropertyGroup>
    <IsAotCompatible>true</IsAotCompatible>
</PropertyGroup>
```

> There's also another option - System.Text.Json already supports source generation and allows turning a JsonNode (tree) into a custom object. I'm unsure how exact the mapping between toml and json is, but converting the parsed toml tree to a jsonnode could be a workaround for some.

Yeah, I would probably prefer to avoid this.

**Comment by lilith** [comment]
Well, I don't think we want to tell downstream libs that it is AOT compatible until we've annotated which APIs are and aren't.

> Yeah, I would probably prefer to avoid this.

Avoid source generation or System.Text.Json integration?

**Comment by xoofx** [comment]
> Avoid source generation or System.Text.Json integration?

`System.Text.Json`

**Comment by xoofx** [comment]
> Well, I don't think we want to tell downstream libs that it is AOT compatible until we've annotated which APIs are and aren't.

Yeah, it was just to suggest that this flag cover the 3 analyzers and he would use it in the end, but probably not during the transition if it takes more than one PR.

**Comment by lilith** [comment]
Of course. I'd say we could leave the warnings on for now?


Tangentially, is there a configuration provider for Tomlyn?
https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.fileconfigurationprovider?view=dotnet-plat-ext-8.0

**Comment by gehongyan** [comment]
Currently, in NativeAOT, I use the source-generated `JsonSerializerContext` to retrieve property field names and pass them through `TomlModelOptions.GetPropertyName`.

```csharp
// A static method for retrieving property name.
static string GetJsonPropertyName(PropertyInfo propertyInfo, JsonSerializerContext jsonSerializerContext)
{
    if (propertyInfo.DeclaringType is not { } declaringType
        || jsonSerializerContext.GetTypeInfo(declaringType) is not { } typeInfo
        || typeInfo.Properties.SingleOrDefault(y => y.AssociatedParameter?.Name == propertyInfo.Name) is not { } property)
        return propertyInfo.Name;
    return property.Name;
}


// Generate the JSON serialzier contexts.
[JsonSerializable(typeof(AppOptions))]
public partial class AppJsonContexts : JsonSerializerContext;

// Pass property names via GetPropertyName 
AppOptions options = Toml.ToModel<AppOptions>(toml, null, new TomlModelOptions
{
    IgnoreMissingProperties = true,
    GetPropertyName = x => GetJsonPropertyName(x, AppJsonSerializerContexts.Default)
});

```
